File: baseunits/modules/Webtoons.pas
unit Webtoons;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, WebsiteModules, uData, uBaseUnit, uDownloadsManager,
  XQueryEngineHTML, httpsendthread, synautil, RegExpr;

implementation

const
  dirurl = '/en/genre';

function GetNameAndLink(const MangaInfo: TMangaInformation;
  const ANames, ALinks: TStringList; const AURL: String;
  const Module: TModuleContainer): Integer;
var
  v: IXQValue;
begin
  Result := NET_PROBLEM;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  if MangaInfo.FHTTP.GET(Module.RootURL + dirurl) then begin
    Result := NO_ERROR;
    with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do
      try
        for v in XPath('//ul[@class="card_lst"]/li/a') do begin
          ALinks.Add(v.toNode.getAttribute('href'));
          ANames.Add(XPathString('div/p[@class="subj"]', v.toNode));
        end;
      finally
        Free;
      end;
  end;
end;

function GetInfo(const MangaInfo: TMangaInformation;
  const AURL: String; const Module: TModuleContainer): Integer;
var
  query: TXQueryEngineHTML;
  p, i: Integer;

  procedure getchapters;
  var
    v: IXQValue;
  begin
    for v in query.XPath('//ul[@id="_listUl"]/li/a') do begin
      MangaInfo.mangaInfo.chapterLinks.Add(v.toNode.getAttribute('href'));
      MangaInfo.mangaInfo.chapterName.Add(query.XPathString('span[@class="subj"]/span', v.toNode));
    end;
  end;

  procedure getp;
  begin
    p := StrToIntDef(SeparateRight(
      query.XPathString('//div[@class="detail_lst"]/div[@class="paginate"]/a[last()]/@href'), '&page='), 1);
  end;

begin
  Result := NET_PROBLEM;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  with MangaInfo.FHTTP, MangaInfo.mangaInfo do begin
    url := FillHost(Module.RootURL, AURL);
    if GET(url) then begin
      Result := NO_ERROR;
      query := TXQueryEngineHTML.Create(Document);
      with query do
        try
          coverLink := XPathString('//meta[@name="twitter:image"]/@content');
          if title = '' then title := XPathString('//div[@class="info"]/h1/text()');
          authors := XPathString('//div[@class="info"]/a/text()');
          genres := XPathString('//div[@class="info"]/h2');
          summary := XPathString('//p[@class="summary"]');
          getchapters;
          getp;
          if p > 1 then begin
            i := 2;
            while (i <= p) and (Thread.IsTerminated = False) do begin
              if GET(url + '&page=' + IntToStr(i)) then begin
                ParseHTML(Document);
                getchapters;
                getp;
              end;
              Inc(i);
            end;
          end;
          InvertStrings([chapterLinks, chapterName]);
          Reset;
          Headers.Values['Referer'] := ' ' + url;
        finally
          query.Free;
        end;
    end;
  end;
end;

function GetPageNumber(const DownloadThread: TDownloadThread;
  const AURL: String; const Module: TModuleContainer): Boolean;
var
  v: IXQValue;
begin
  Result := False;
  if DownloadThread = nil then Exit;
  with DownloadThread.FHTTP, DownloadThread.Task.Container do begin
    PageLinks.Clear;
    PageNumber := 0;
    if GET(FillHost(Module.RootURL, AURL)) then begin
      Result := True;
      with TXQueryEngineHTML.Create(Document) do
        try
          with TRegExpr.Create do
            try
              Expression := '[\?\&]type=q\d+';
              for v in XPath('//div[@id="_imageList"]/img[@class="_images"]/@data-url') do
                PageLinks.Add(Replace(v.toString, '', False));
            finally
              Free;
            end;
        finally
          Free;
        end;
    end;
  end;
end;

function BeforeDownloadImage(const DownloadThread: TDownloadThread;
  var AURL: String; const Module: TModuleContainer): Boolean;
begin
  Result := False;
  if DownloadThread = nil then Exit;
  with DownloadThread.Task.Container do
    if CurrentDownloadChapterPtr < ChapterLinks.Count then begin
      DownloadThread.FHTTP.Headers.Values['Referer'] :=
        ' ' + FillHost(Module.RootURL, ChapterLinks[CurrentDownloadChapterPtr]);
      Result := True;
    end;
end;

procedure RegisterModule;
begin
  with AddModule do
  begin
    Website := 'Webtoons';
    RootURL := 'http://www.webtoons.com';
    Category := 'English';
    OnGetNameAndLink := @GetNameAndLink;
    OnGetInfo := @GetInfo;
    OnGetPageNumber := @GetPageNumber;
    OnBeforeDownloadImage := @BeforeDownloadImage;
  end;
end;

initialization
  RegisterModule;

end.

File: lua/modules/WebToons.lua
local langs = {
  ["en"] = "English",
  ["id"] = "Indonesian",
  ["zh-hant"] = "Chinese",
  ["th"] = "Thai"
}

function getinfo()
  mangainfo.url = mangainfo.url:gsub('(.*)&page=.*', '%1')
  http.cookies.values['ageGatePass'] = 'True'
  if http.get(mangainfo.url) then
    x=TXQuery.Create(http.document)
    mangainfo.title = x.xpathstring('//meta[@property="og:title"]/@content')
    mangainfo.coverlink=x.xpathstring('//meta[@name="twitter:image"]/@content')
    mangainfo.authors=x.xpathstring('//div[@class="info"]//span[@class="author"]')
    mangainfo.genres=x.xpathstring('//div[@class="info"]/h2')
    if mangainfo.genres == '' then
      mangainfo.genres=x.xpathstring('//div[@class="info challenge"]/p')
    end
    mangainfo.summary=x.xpathstring('//p[@class="summary"]')
    local pages = 1
    local p = 1
    while p <= pages do
      if p > 1 then
        if http.get(mangainfo.url .. '&page=' .. tostring(p)) then
          x=TXQuery.Create(http.document)
        else
          break
        end
      end
      if p == pages then
        local pg = x.xpathstring('//div[@class="detail_lst"]/div[@class="paginate"]/a[last()]/substring-after(@href, "&page=")')
        if pg ~= '' then pages = tonumber(pg) end
      end
      local v=x.xpath('//div[@class="detail_lst"]/ul/li/a')
      for i=1,v.count do
        local v1=v.get(i)
        mangainfo.chapterlinks.add(v1.getAttribute('href'))
        mangainfo.chapternames.add(x.xpathstring('.//span[@class="subj"]/span', v1))
      end
      p = p + 1
    end
    InvertStrings(mangainfo.chapterlinks,mangainfo.chapternames)
    return no_error
  else
    return net_problem
  end
end

function getpagenumber()
  task.pagenumber=0
  task.pagelinks.clear()
  url = url:gsub('(.*)&page=.*', '%1')
  http.Cookies.Values['ageGatePass'] = 'True'
  if http.get(MaybeFillHost(module.rooturl,url)) then
    TXQuery.Create(http.Document).xpathstringall('//div[@id="_imageList"]/img[@class="_images"]/@data-url', task.pagelinks)
    return true
  else
    return false
  end
end

function getnameandlink()
  local selectedLang = module.getoption('lualang')
  local l = langs
  if selectedLang > 0 then
    l = {[findlang(selectedLang)] = ""}
  end
  for key, value in pairs(l) do
    local dirurl = key..'/genre'
    if http.get(module.rooturl..dirurl) then
      local x=TXQuery.Create(http.Document)
	    local v = x.xpath('//div[@class="card_wrap genre"]/ul/li/a')
	    for i = 1, v.count do
        local v1 = v.get(i)
	      names.Add(x.xpathstring('.//div[@class="Info"]//p[@class="subj"]', v1)..' ['..key..']');
		    links.Add(v1.getAttribute('href'));
	    end
	    return no_error
    else
      return net_problem
    end
  end
  if module.getoption('luaincludechallengetitles') then
    getnameandlinkforchallenge()
  end
end

function getnameandlinkforchallenge()
  local selectedLang = module.getoption('lualang')
  local l = langs
  if selectedLang > 0 then
    l = {[findlang(selectedLang)] = ""}
  end
  for key, value in pairs(l) do
    local dirurl = key..'/challenge/list?genreTab=ALL&sortOrder=UPDATE'
    if http.get(module.rooturl..dirurl) then
      local x=TXQuery.Create(http.Document)
      
      
      local pages = 1
      local p = 1
      while p <= pages do
        if p > 1 then
          if http.get(module.rooturl..dirurl..'&page='..tostring(p)) then
            x=TXQuery.Create(http.document)
          else
            break
          end
        end
        if p == pages then
          local pg = x.xpathstring('//div[@class="paginate"]/a[last()]/substring-after(@href, "&page=")')
          if pg ~= '' then pages = tonumber(pg) end
        end
	      local v = x.xpath('//div[@class="challenge_cont_area"]/div[contains(@class,"challenge_lst")]/ul/li/a[contains(@class,"challenge_item")]')
	      for i = 1, v.count do
          local v1 = v.get(i)
	        names.Add(x.xpathstring('./p[@class="subj"]', v1)..' ['..key..']');
		      links.Add(v1.getAttribute('href'));
	      end
        p = p + 1
      end
      
      
	    return no_error
    else
      return net_problem
    end
  end
end

function BeforeDownloadImage()
  http.headers.values['Referer'] = url
  return true
end

function getlang(lang)
  if langs[lang] ~= nil then
    return langs[lang]
  else
    return 'Unknown'
  end
end

function getlanglist()
  local t = {}
  for k, v in pairs(langs) do table.insert(t, v); end
  table.sort(t)
  return t
end

function findlang(lang)
  local t = getlanglist()
  for i, v in ipairs(t) do
    if i == lang then
      lang = v
      break
    end
  end
  for k, v in pairs(langs) do
    if v == lang then return k; end
  end
  return nil
end

function Init()
  local m=NewModule()
  m.category='English'
  m.website='WebToons'
  m.rooturl='https://www.webtoons.com/'
  m.lastupdated='April 14, 2019'
  m.ongetinfo='getinfo'
  m.ongetpagenumber='getpagenumber'
  m.ongetnameandlink='getnameandlink'
  m.OnBeforeDownloadImage = 'BeforeDownloadImage'
  
  m.addoptioncheckbox('luaincludechallengetitles', 'Include manga titles from WebToons Challenge (takes very very long to create manga list!):', false)
  
  local items = 'All'
  local t = getlanglist()
  for k, v in ipairs(t) do items = items .. '\r\n' .. v; end
  m.addoptioncombobox('lualang', 'Language:', items, 2)
end
